src/lem/constrainer.rs

```diff
@@ -81,103 +150,517 @@ impl LEM {
+                        hash_slots.hash3_stacks.implies_stack.push((
+                            concrete_path.clone(),
+                            slots.hash3_idx,
+                            hash,
+                            tag,
+                        ));
+                    }
+                }
+            }
+            AllocHashPreimage::A4(i0, i1, i2, i3) => {
+                match path_kind {
+                    PathKind::Concrete => {
+                        hash_slots
+                            .hash4_stacks
+                            .enforce_stack
+                            .push((slots.hash4_idx, hash, tag));
+                        hash_slots
+                            .hash4_alloc
+                            .push((slots.hash4_idx, i0, i1, i2, i3))
+                    }
+                    PathKind::MaybeConcrete(concrete_path) => {
+                        // if concrete_path is true, push to slots
+                        if is_concrete_path {
+                            hash_slots
+                                .hash4_alloc
+                                .push((slots.hash4_idx, i0, i1, i2, i3));
+                            // only once per path
+                        }
+                        // concrete path implies alloc_tgt has the same value as in the current slot
+                        hash_slots.hash4_stacks.implies_stack.push((
+                            concrete_path.clone(),
+                            slots.hash4_idx,
+                            hash,
+                            tag,
+                        ));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Here we use the implies/enforce logic to contrain tag and hash values.
+    /// We pop stacked elements in order to get virtual path information. When
+    /// many branches are possible we will use the `implies_equal` gadget to
+    /// ensure only concrete paths are implied. Otherwise, only one path exists,
+    /// therefore it must be enforced.
+    /// Hence we use two distinct stacks, one for implications and another one
+    /// for enforcements.
+    /// Beyond that, we need to deal with the possibility of receiving no tag,
+    /// since Unhash operations don't require constraining tags.
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        alloc_ptrs: &HashMap<&String, AllocatedPtr<F>>,
+        implies_stack: &Vec<(Boolean, usize, MetaPtr, Option<Tag>)>,
+        enforce_stack: &Vec<(usize, MetaPtr, Option<Tag>)>,
+        hash_slots: &HashMap<usize, AllocatedNum<F>>,
+        alloc_manager: &mut AllocationManager<F>,
+    ) -> Result<()> {
+        // Create hash implications
+        for (concrete_path, slot, tgt, tag) in implies_stack {
+            // get alloc_tgt from tgt
+            let Some(alloc_tgt) = alloc_ptrs.get(tgt.name()) else {
+                bail!("{} not allocated", tgt.name());
```

Comment:
What does it *mean* for this error to occur? Does this mean:
- That the concrete LEM being constrained has a bug in it?
- That the constraint algorithm and/or implementation has a bug in it?
- That there is a bug in the program being evaluated by the LEM?
- Something else?
-

Reply:
First option. It means that the LEM in question makes use of some meta pointer that hasn't been defined before.

At least that's what `... not allocated` errors usually mean in the constrainer code. But @emmorais was the one who wrote this part so he might know better.

Reply:
I think we could use some typed errors that begin to broadly distinguish things like those I enumerated above.

Reply:
Maybe it's the case that 'first option' is the only kind of error that happens anywhere, in which case, I guess it's not pressing.

Reply:
Yes, it means the LEM programmer did something like `Hash2(non-existing-var-name, ...)`.

Reply:
For clarity, are you saying that *all* errors produced in the LEM code are the result of bad LEM programs? That's what I meant in the comment directly above.

Reply:
I don't know, that would be a strong claim and would require a more careful investigation